tests/unit/sidecars/secrets_fetcher_tests.py

```diff
@@ -0,0 +1,224 @@
+                    "secret/one": "SECRET/ONE",
+                    "secret/two": "SECRET/TWO",
+                    "secret/three": "SECRET/THREE",
+                },
+                "vault": {
+                    "token": "token",
+                    "url": "https://vault.example.com:8200/",
+                },
+                "vault_token": "token",
+            },
+        )
+
+
+class BadJSONTests(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        spec = {
+            "vault": {
+                "url": config.DefaultFromEnv(config.String, "BASEPLATE_DEFAULT_VAULT_URL"),
+                "role": config.String,
+                "auth_type": config.Optional(
+                    config.OneOf(**secrets_fetcher.VaultClientFactory.auth_types()),
+                    default=secrets_fetcher.VaultClientFactory.auth_types()["aws"],
+                ),
+                "mount_point": config.DefaultFromEnv(
+                    config.String, "BASEPLATE_VAULT_MOUNT_POINT", fallback="aws-ec2"
+                ),
+            },
+            "output": {
+                "path": config.Optional(config.String, default="/var/local/secrets.json"),
+                "owner": config.Optional(config.UnixUser, default=0),
+                "group": config.Optional(config.UnixGroup, default=0),
+                "mode": config.Optional(config.Integer(base=8), default=0o400),  # type: ignore
+            },
+            "secrets": config.Optional(config.TupleOf(config.String), default=[]),
+            "callback": config.Optional(config.String),
+        }
+
+        parser = configparser.RawConfigParser()
+        with io.StringIO(configini) as f:
+            parser.read_file(f)
+        fetcher_config = dict(parser.items("secret-fetcher"))
+
+        cls.cfg = config.parse_config(fetcher_config, spec)
+
+    def setUp(self):
+        self.setUpPyfakefs()
+        self.fake_fs().create_file(
+            "/var/local/secrets.json",
+            contents="initial contents should remain unchanged",
+        )
+
+        cfg = self.cfg
+        now = datetime.datetime.now(UTC)
+
+        with unittest.mock.patch("baseplate.sidecars.secrets_fetcher.VaultClientFactory") as mock:
+            instance = mock.return_value
+            instance.get_client.return_value = FakeBadVaultClient(token_expiration=now)
+            f = secrets_fetcher.VaultClientFactory(
+                cfg.vault.url, cfg.vault.role, cfg.vault.auth_type, cfg.vault.mount_point
+            )
+            with self.assertRaises(TypeError):
+                secrets_fetcher.fetch_secrets(cfg, f)
+
+    def test_temporary_file_is_not_deleted(self):
+        p = pathlib.Path("/var/local/secrets.json.tmp")
+        self.assertTrue(p.exists())
+
+    def test_temporary_file_is_partially_written(self):
+        p = pathlib.Path("/var/local/secrets.json.tmp")
+        text = p.read_text()
+        self.assertEqual(text, """{\n  "secrets": {\n    "secret/one": """)
```

Comment:
should this assert that the file contents either don't include a valid token or that the contents are not valid json rather than matching the malformed content in case the malformed content changes in the future.

Reply:
This treatment showcases the general issue and how it precisely manifests in this failure mode. I think the check for invalid JSON is a weaker assertion and the maintenance burden with content changes doesn't outweigh that.

I don't think this partially written file side effect is desirable or beneficial and could be changed. I'd previously done that but didn't want to submit a PR against code that is not tested -- so maybe this test will go away soon anyway.